[serve] Add documentation for `max_queued_requests` (#43501)
ray-project/ray: doc/source/serve/production-guide/best-practices.md

:end-before: __production_code_end__
:language: python
```

## Load shedding

> When a request is sent to a cluster, it is first received by the Serve proxy, which then forwards it to a replica for handling using a {mod}`DeploymentHandle <ray.serve.handle.DeploymentHandle>`.

Reviewer: ```suggestion
When a request is sent to a cluster, it's first received by the Serve proxy, which then forwards it to a replica for handling using a {mod}`DeploymentHandle <ray.serve.handle.DeploymentHandle>`.
```

> Replicas can handle up to a configurable number of requests at a time (configured using the `max_ongoing_requests` option).

Reviewer: ```suggestion
Replicas can handle up to a configurable number of requests at a time. Configure the number using the `max_ongoing_requests` option.
```

> If all replicas are busy and cannot accept more requests, the request will be queued in the {mod}`DeploymentHandle <ray.serve.handle.DeploymentHandle>` until one becomes available.

Reviewer: ```suggestion
If all replicas are busy and cannot accept more requests, the request is queued in the {mod}`DeploymentHandle <ray.serve.handle.DeploymentHandle>` until one becomes available.
```

Under heavy load, {mod}`DeploymentHandle <ray.serve.handle.DeploymentHandle>` queues can grow and cause high tail latency and excessive load on the system.
To avoid instability, it's often preferable to intentionally reject some requests to avoid these queues growing indefinitely.
This technique is called "load shedding," and it allows the system to gracefully handle excessive load without spiking tail latencies or overloading components to the point of failure.

You can configure load shedding for your Serve deployments using the `max_queued_requests` parameter to the {mod}`@serve.deployment <ray.serve.deployment>` decorator.

> This controls the maximum number of requests that can be queued in each {mod}`DeploymentHandle <ray.serve.handle.DeploymentHandle>` (including each Serve proxy).

Reviewer: ```suggestion
This parameter controls the maximum number of requests that Serve queues in each {mod}`DeploymentHandle <ray.serve.handle.DeploymentHandle>`, including each Serve proxy.
```

> Once the limit is reached, enqueueing any new requests will immediately raise a {mod}`BackPressureError <ray.serve.exceptions.BackPressureError>`.

Reviewer: ```suggestion
Once the limit is reached, enqueueing any new requests immediately raises a {mod}`BackPressureError <ray.serve.exceptions.BackPressureError>`.
```

> Once the limit is reached, enqueueing any new requests will immediately raise a {mod}`BackPressureError <ray.serve.exceptions.BackPressureError>` (TODO: API ref for exception).

Reviewer 2: Just need to remove this TODO 🙃

Author: ah hah!

> HTTP requests will return a `503` status code (service unavailable).

Reviewer: ```suggestion
HTTP requests return a `503` status code, meaning service unavailable.
```

The following example defines a deployment that emulates slow request handling and has `max_concurrent_queries` and `max_queued_requests` configured.

```{literalinclude} ../doc_code/load_shedding.py
:start-after: __example_deployment_start__
:end-before: __example_deployment_end__
:language: python
```

To test the behavior, send HTTP requests in parallel to emulate multiple clients.

> `max_concurrent_queries` + `max_queued_requests` requests will be accepted, and further requests will be rejected with a `503` (service unavailable) status.

Reviewer: ```suggestion
Serve accepts `max_concurrent_queries` and `max_queued_requests` requests, and rejects further requests with a `503`, or service unavailable, status.
```
